alea/include/alps/alea/internal/format.hpp

```diff
@@ -0,0 +1,190 @@
+ *         get_format(std::cerr, TERSE) = verb;
+ *     }
+ *     // ...
+ *     std::ostream &operator<<(std::ostream &str, const my_type &obj) {
+ *         if (get_format(std::cerr, TERSE) == VERBOSE) {
+ *             // ...
+ *         }
+ *     }
+ *
+ * @see format_registry
+ */
+template <typename T>
+T &get_format(std::ios_base &stream, T initial_value = T())
+{
+    return format_registry<T>::get(stream, initial_value);
+}
+
+/**
+ * Allows the addition of custom formatting state to I/O stream objects.
+ *
+ * This class allows to store one instance of type `T` with each `std::ios_base`
+ * instance via the `get()` method.  This in turn can be used to implement
+ * user-defined I/O manipulators that modify these format flags and are then
+ * used by stream operations with user-defined objects.
+ *
+ * @see get_format()
+ */
+template <typename T>
+class format_registry
+{
+public:
+    /** Returns reference to format object for stream, creating it if needed. */
+    static T &get(std::ios_base &stream, T init = T())
+    {
+        T *&format = get_format(stream);
+
+        // if format is not there, we should add it
+        if (format == nullptr) {
+            format = new T(init);     // must be copy constructible
+            stream.register_callback(callback, get_xindex());
+        }
+        return *format;
+    }
+
+    /** Removes format object from stream */
+    static void remove(std::ios_base &stream)
+    {
+        callback(std::ios::erase_event, stream, get_xindex());
+    }
+
+protected:
+    /** Manages lifecycle using the std::ios_base call-back mechanism */
+    static void callback(std::ios::event event, std::ios_base &stream,
+                         int xindex)
+    {
+        if (xindex != get_xindex()) { assert(false); }
+        T *&self = get_format(stream);
+
+        switch (event) {
+        case std::ios::erase_event:
+            // this will be called in destructor, so we MUST NOT throw
+            // exceptions
+            if (self != nullptr) {
+                delete self;
+                self = nullptr;  // to be sure
+            }
+            break;
+
+        case std::ios::copyfmt_event:
+            // copyfmt first triggers an erase_event, which we use to clean up
+            // the memory, then does a shallow copy, and then triggers this
+            // event.  This means that self here will point to the original
+            // instance.
+            self = new T(*self);
+            break;
+
+        case std::ios::imbue_event:
+            // FIXME
+            break;
+        }
+    }
+
+    static T *& get_format(std::ios_base &stream)
+    {
+        return reinterpret_cast<T *&>(stream.pword(get_xindex()));
+    }
+
+    static int get_xindex()
+    {
+        // guaranteed to be distinct since for different iomanip's since their
+        // T will be different.
+        static int xindex = std::ios_base::xalloc();
```

Comment:
> This call is a bit unsettling...
> http://en.cppreference.com/w/cpp/io/ios_base/xalloc
> This function is thread-safe; concurrent access by multiple threads does not result in a data race

I don't know if it is that much of a problem in practice -- any implementation I've seen of xalloc is thread-safe because this thing has been around since C++98 and otherwise all uses of the STL in the multi-threaded context would immediately blow up.